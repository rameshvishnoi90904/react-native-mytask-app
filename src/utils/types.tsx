export type task = {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export type dispatchActionType = {
  type: string;
  data: task | null;
  searchInput: string;
}

export type NavigationScreenPropsType = {
  Home: undefined;
  TodoScreen: {item: task | null };
  LandingScreen: {};
};
