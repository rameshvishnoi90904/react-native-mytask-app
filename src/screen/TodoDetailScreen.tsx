import React from "react"

const TodoDetailScreen = () => {
    return (
        <></>
    )
}
export default TodoDetailScreen;