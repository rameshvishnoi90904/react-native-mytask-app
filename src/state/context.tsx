
import { createContext } from 'react';
import { dispatchActionType, task } from '../utils/types';

export type AppContextType = {
    tasks: task[];
    dispatch: (action: dispatchActionType) => void;
}

export const AppContext = createContext<AppContextType | null>(null);