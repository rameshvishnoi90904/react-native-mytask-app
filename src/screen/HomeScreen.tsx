import { Button, FlatList, Platform, StyleSheet, Text, TextInput, Touchable, TouchableOpacity, View } from "react-native"
import React, { useContext, useEffect, useState } from "react";

import type { NativeStackScreenProps } from '@react-navigation/native-stack';

import { AppContext, AppContextType } from "../state/context";
import { NavigationScreenPropsType, task } from "../utils/types";
import TodoItem from "../components/TodoItem";
type Props = NativeStackScreenProps<NavigationScreenPropsType, 'Home'>;

const HomeScreen = ({navigation, route}: Props) => {
    const navigateToDetailScreen = (item: task) => {
        navigation.navigate('TodoScreen', {item: item});
    }
    const {tasks, dispatch} = useContext(AppContext) as AppContextType;
    const [searchInput, setSearchInput] = useState('')

    useEffect(() => {
        dispatch({type: 'search', data: null, searchInput: searchInput.toLowerCase()})
    },[searchInput])

    const addNewTask = () => {
        navigation.navigate('TodoScreen', {item: {title: '', body: '', id: 0, userId: 0}});
    }

    const emptySearchResult = () => {
        if (searchInput.length > 1) {
            return (
                <View style={{flex: 1}}>
                    <Text>No Tasks matching your search</Text>
                </View>
            )
        } else {
            return (
                <View style={{flex: 1}}>
                    <Text>No Tasks pending</Text>
                </View>
            )
        }
    }
    return (
        <View style={styles.container}>
            <View style={styles.searchContainer}>
                <TextInput style={styles.searchInput} placeholder="Search here" value={searchInput} onChangeText={setSearchInput}/>
            </View>
            <FlatList 
                data={tasks}
                testID="list-view"
                ListEmptyComponent={emptySearchResult()}
                renderItem={({item}) => <TodoItem item={item} navigateToDetailScreen={navigateToDetailScreen}/>}
                keyExtractor={item => String(item.id)}
            />
            <TouchableOpacity testID="add-new-task" style={styles.floatingButton} onPress={addNewTask}>
                <Text style={styles.addButton}>{'+'}</Text>
            </TouchableOpacity>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        paddingTop: 12,
        paddingHorizontal: 16,
        flex: 1
    },
    
    searchContainer: {
        marginBottom: 12,
        borderRadius: 12,
        borderWidth: 1,
        borderColor: '#DCDCDC',
        ...Platform.select({
            ios: {
                paddingHorizontal: 8,
                paddingVertical: 10,
            },
            android: {
            },
        })
    },
    searchInput: {
        borderRadius: 12
    },
    floatingButton: {
        elevation: 4,
        position: "absolute",
        height: 56,
        width: 56,
        borderRadius: 28,
        backgroundColor: '#FB8DA0',
        alignItems: 'center',
        justifyContent: 'center',
        textAlign: 'center',
        bottom: 50,
        right: 30
    },
    addButton: {
        color: 'white',
        fontSize: 40,
    }
})
  

export default HomeScreen;