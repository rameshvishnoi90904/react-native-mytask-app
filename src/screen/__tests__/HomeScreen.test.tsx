import { render, screen, fireEvent } from '@testing-library/react-native';
import {it, jest, expect} from '@jest/globals';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { AppContext } from '../../state/context';
import { NavigationScreenPropsType, dispatchActionType, task } from '../../utils/types';
import HomeScreen from '../HomeScreen';
import TodoDetailScreen from '../TodoDetailScreen';

const Stack = createNativeStackNavigator<NavigationScreenPropsType>();


it('it renders correctly', () => {
    const mockTasks: Array<task> = [];
    const mockDispatch = jest.fn()

    const tree = render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="Home" component={HomeScreen}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    ).toJSON()

    expect(tree).toMatchSnapshot();
})


it('it renders correctly with n task', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="Home" component={HomeScreen}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )

    expect(screen.getByTestId('list-view').props.data).toStrictEqual(mockTasks)
})

it('adds new task', () => {
    const mockTasks: Array<task> = [];
    const mockDispatch = jest.fn()

    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="Home" component={HomeScreen}/>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )

    fireEvent.press(screen.getByTestId('add-new-task'))
    expect(screen.getByText('Body')).toBeTruthy()
})



it('open 1st task', () => {
    const mockTasks: Array<task> = [
        {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
        {id: 98, userId: 99, title: "another sample", body: "another sample body"}
    ];
    const mockDispatch = jest.fn()

    render(
        <AppContext.Provider value={{tasks: mockTasks, dispatch: mockDispatch}}>
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="Home" component={HomeScreen}/>
                    <Stack.Screen name="TodoScreen" component={TodoDetailScreen}/>
                </Stack.Navigator>
            </NavigationContainer>
        </AppContext.Provider>
    )


    fireEvent.press(screen.getByTestId(`todoItem-${mockTasks[0].id}`))
    expect(screen.getByText(mockTasks[0].title)).toBeTruthy();
})