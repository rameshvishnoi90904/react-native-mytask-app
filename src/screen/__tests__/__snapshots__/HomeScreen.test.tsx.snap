// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "flex": 1,
                "paddingHorizontal": 16,
                "paddingTop": 12,
              }
            }
          >
            <View
              style={
                {
                  "borderColor": "#DCDCDC",
                  "borderRadius": 12,
                  "borderWidth": 1,
                  "marginBottom": 12,
                  "paddingHorizontal": 8,
                  "paddingVertical": 10,
                }
              }
            >
              <TextInput
                onChangeText={[Function]}
                placeholder="Search here"
                style={
                  {
                    "borderRadius": 12,
                  }
                }
                value=""
              />
            </View>
            <RCTScrollView
              ListEmptyComponent={
                <View
                  style={
                    {
                      "flex": 1,
                    }
                  }
                >
                  <Text>
                    No Tasks pending
                  </Text>
                </View>
              }
              data={[]}
              getItem={[Function]}
              getItemCount={[Function]}
              keyExtractor={[Function]}
              onContentSizeChange={[Function]}
              onLayout={[Function]}
              onMomentumScrollBegin={[Function]}
              onMomentumScrollEnd={[Function]}
              onScroll={[Function]}
              onScrollBeginDrag={[Function]}
              onScrollEndDrag={[Function]}
              removeClippedSubviews={false}
              renderItem={[Function]}
              scrollEventThrottle={50}
              stickyHeaderIndices={[]}
              testID="list-view"
              viewabilityConfigCallbackPairs={[]}
            >
              <View>
                <View
                  onLayout={[Function]}
                  style={
                    {
                      "flex": 1,
                    }
                  }
                >
                  <Text>
                    No Tasks pending
                  </Text>
                </View>
              </View>
            </RCTScrollView>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FB8DA0",
                  "borderRadius": 28,
                  "bottom": 50,
                  "elevation": 4,
                  "height": 56,
                  "justifyContent": "center",
                  "opacity": 1,
                  "position": "absolute",
                  "right": 30,
                  "textAlign": "center",
                  "width": 56,
                }
              }
              testID="add-new-task"
            >
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 40,
                  }
                }
              >
                +
              </Text>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Home"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
